### PR TITLE
Add mast3r slam adapter

### PR DESCRIPTION
**Pull request by christopherkirschner**

### CODE REVIEW

**Comment by JanDuchscherer104** [comment]
Request Copilot summary and review first!

**Comment by JanDuchscherer104** [comment]
@copilot+claude-sonnet-4.6 please review this PR. identify all issues and action items. give a summary of all changes!

File: src/prml_vslam/methods/mast3r/adapter.py [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+            img_shape = mast3r_frame.img_shape
+            if img_shape is None:
+                return None, None, 0
+            shape_vals = img_shape.flatten().tolist()
+            if len(shape_vals) < 2:
+                return None, None, 0
+            h, w = int(shape_vals[0]), int(shape_vals[1])
+
+            # Pose X_canon (camera frame) into world frame for consistent Rerun logging.
+            points_world = mast3r_frame.T_WC.act(x_canon).detach().cpu().numpy().astype(np.float32)
```
Comment (JanDuchscherer104) [inline src/prml_vslam/methods/mast3r/adapter.py]: TODO(01): _extract_keyframe_visuals emitts world-space coords. SlamUpdate must contain point-maps in the local camera frame, with `slam_update.pose` being the corresponding T_WC SE3 TF.

File: src/prml_vslam/methods/mast3r/adapter.py [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+            # Start the backend thread now that dimensions are known.
+            self._backend_stop.clear()
+            self._backend_thread = threading.Thread(
+                target=self._backend_loop, name="mast3r-backend", daemon=True
+            )
+            self._backend_thread.start()
+            self._console.info(f"MASt3R initialized with dynamic size: {self._h}x{self._w}")
+
+        import torch  # noqa: PLC0415
+        from mast3r_slam.frame import Mode, create_frame  # noqa: PLC0415
```
Comment (JanDuchscherer104) [inline src/prml_vslam/methods/mast3r/adapter.py]: Remove imports from step(), I know, ViSTA on main does the same.

File: src/prml_vslam/methods/mast3r/adapter.py [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+    def step(self, frame: FramePacket) -> None:
+        """Feed one frame to MASt3R and emit an incremental SlamUpdate."""
+        if frame.rgb is None:
+            self._emit_empty_update(frame)
+            return
+        
+        if self._keyframes is None:
+            from mast3r_slam.mast3r_utils import resize_img
+            from mast3r_slam.frame import SharedKeyframes, SharedStates
+            from mast3r_slam.tracker import FrameTracker
```
Comment (JanDuchscherer104) [inline src/prml_vslam/methods/mast3r/adapter.py]: remove imports from step()

File: src/prml_vslam/methods/mast3r/adapter.py [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+        import torch  # noqa: PLC0415
+        from mast3r_slam.frame import Mode, create_frame  # noqa: PLC0415
+
+        internal_idx = self._source_frame_count
+        self._timestamps_s.append(frame.timestamp_ns / 1e9)
+
+        # If we're configured for use_calib and we haven't set K yet, grab it
+        # from the packet. We compute K_frame on first use (scaled intrinsics
+        # for the resized MASt3R image).
+        self._maybe_set_intrinsics(frame)
```
[the remark this answers is not in the document]
Comment (JanDuchscherer104) [reply src/prml_vslam/methods/mast3r/adapter.py]: I don't fully understand this issue, but GPT flagged this as well!

```
1. Calibration race: backend thread starts before _K is set

This is the biggest correctness blocker. In step(), the backend thread is started immediately after creating SharedKeyframes / SharedStates, and only later does the frame path call _maybe_set_intrinsics(frame). The backend thread constructs FactorGraph(self._model, self._keyframes, self._K, self._device) in _backend_loop, so with use_calib=True, it can observe self._K is None even if the incoming FramePacket has intrinsics. 

Action: move _maybe_set_intrinsics(frame) before starting the backend thread, or start the backend loop lazily only after the first calibrated frame has populated _K. Add a unit-level regression using fake upstream modules where use_calib=True and the backend loop asserts that K is not None.
```

**Review by JanDuchscherer104 — CHANGES_REQUESTED** [review]
ATM I think that I should merge my branch first and then yours on top? You're currently doing plenty of work inside adapter.py that will no longer be required.

So definitely fix the fame consistency issue. make sure that the outputted frames are in RDF (Right Down Forward), as well as the race condition that has been flagged by copilot and gpt.

**Comment by JanDuchscherer104** [comment]
[GPT-5.5 Pro Review](https://chatgpt.com/share/69eb7b6c-a658-8396-ad2c-2ac3e9075644):

## Current blockers / issues with file + line numbers

| Severity               |                                                                        File / lines | Issue                                                                                                                                                                                                                                        | Action                                                                                                                                     |
| ---------------------- | ----------------------------------------------------------------------------------: | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
| **Blocker**            |      `src/prml_vslam/methods/mast3r/adapter.py:L182-L199`, `L349-L356`, `L477-L523` | **Calibration race.** Backend thread starts at `L182-L187`; `_maybe_set_intrinsics()` runs later at `L196-L199`; backend loop constructs `FactorGraph(..., self._K, ...)` at `L356`. If `use_calib=True`, backend may see `self._K is None`. | Call `_maybe_set_intrinsics(frame)` before starting the backend thread, or lazily start backend after first calibrated frame has set `_K`. |
| **Blocker**            |                                                              `adapter.py:L279-L289` | **`close()` assumes `_states` exists.** If no RGB frame was accepted, `_states` is `None`, and `self._states.set_mode(...)` crashes.                                                                                                         | Guard zero-frame/no-init case and raise a clear “no frames processed” error.                                                               |
| **Blocker**            |                                                              `adapter.py:L429-L431` | **Backend thread errors are swallowed.** `_backend_loop()` logs “continuing” and sleeps. The run can finish with invalid state.                                                                                                              | Store `_backend_error`; raise it from `step()`, `try_get_updates()`, or `close()`.                                                         |
| **Blocker**            |                                                              `adapter.py:L735-L750` | **Offline path does not clean up on failure.** If `cv2.imread`, `session.step`, or inference fails, `session.close()` is skipped and the backend thread can remain alive.                                                                    | Wrap offline loop in `try/finally`; call `session.close()` or `session.abort()` on failure.                                                |
| **Blocker**            |                                                              `adapter.py:L747-L749` | **Offline timestamps ignore `SequenceManifest.timestamps_path`.** Every frame gets synthetic 30 FPS timestamps. That can break trajectory evaluation.                                                                                        | Load normalized timestamps from `SequenceManifest.timestamps_path`; use 30 FPS only as fallback.                                           |
| **Blocker / semantic** | `adapter.py:L597-L611`; current protocol says `SlamUpdate.pointmap` is camera-local | **Pointmap frame mismatch.** Adapter transforms `X_canon` to world points before putting it into `SlamUpdate.pointmap`, but current DTO docs describe a camera-local pointmap.                                                               | Either keep pointmap camera-local, or add an explicit coordinate-frame field before emitting world-space points.                           |
| **High**               |                                                              `adapter.py:L846-L848` | **PLY colors are dropped when Open3D is installed.** Reads point cloud, extracts only `points`, then writes through `write_point_cloud_ply(...)`.                                                                                            | Preserve colors or copy native PLY byte-for-byte consistently.                                                                             |
| **High**               |                                                              `adapter.py:L855-L858` | **Same PLY is assigned as sparse and dense depending on policy.** Artifact semantics are unclear.                                                                                                                                            | Pick one semantic default, likely `dense_points_ply`, unless MASt3R exposes a true sparse output.                                          |
| **Medium**             |                                                 `adapter.py:L129-L130`, `L166-L188` | `_initialize()` docstring says it builds shared state and starts backend thread, but thread actually starts lazily in `step()`.                                                                                                              | Update docstring.                                                                                                                          |
| **Medium**             |                                                              `adapter.py:L597-L607` | Preview shape is not normalized. `uimg` layout might not be guaranteed `H×W×3 uint8`.                                                                                                                                                        | Add helper that guarantees RGB `uint8` image shape.                                                                                        |
| **Medium**             |                                                              `adapter.py:L629-L634` | `_inject_sys_path()` only inserts the MASt3R root and comments that editable installs handle third-party dirs. This is fragile in non-editable envs.                                                                                         | Either require editable installs strictly or add robust third-party path handling / better prerequisite checks.                            |
| **Medium**             |                                  `src/prml_vslam/pipeline/run_service.py:L179-L186` | MASt3R backend construction is wired directly into old `_default_slam_backend_factory`. This will be rewritten by the big refactor.                                                                                                          | If merging after refactor, do not preserve this old wiring; port via method-owned backend/stage plugin.                                    |
| **Resolved / OK**      |                            `src/prml_vslam/pipeline/contracts/request.py:L152-L159` | Earlier Copilot bug about `if` vs `elif` appears fixed: VISTA and MASt3R coercion are now one `if/elif/else` chain.                                                                                                                          | No action.                                                                                                                                 |
| **Resolved / OK**      |                               `src/prml_vslam/app/pages/pipeline.py` current branch | Earlier disabled triple-quoted MASt3R help branch appears fixed in current raw head.                                                                                                                                                         | No action unless patch view is stale.                                                                                                      |

@copilot+gpt-5.4 Review each of the identified issues and respective action-itmes and add in-file review comments if you agree with the findings.

**Review by JanDuchscherer104 — APPROVED** [review]

**Comment by JanDuchscherer104** [comment]
@copilot give TLDR status report on all remaining issues (as pointed out in previous code review comments) as well as any new findings.

Always give #L and concrete functional error description followed by proposed patch.
